### PR TITLE
Feature/chained location.near search

### PR DESCRIPTION
**Pull request by jkiddo**
Credits goes to @nigtrifork

### CODE REVIEW

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/UpliftedChainedNearSearchTest.java [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+import java.util.Collections;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+
+public class UpliftedChainedNearSearchTest extends BaseResourceProviderR4Test {
```
Comment (tadgh) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/UpliftedChainedNearSearchTest.java:32]: issue: FWIW based on my read of this, this is not an uplifted chain, just a regular chain. Please do let me know if I'm misunderstanding.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java [hunk 2]
```diff
@@ -378,13 +379,31 @@ public void addSortOnResourceLink(
...
+						IQueryParameterType locationParam = params.get(0).get(0);
+						final SpecialParam specialParam = new SpecialParam().setValue(locationParam.getValueAsQueryToken(myFhirContext));
+						ParsedLocationParam location = ParsedLocationParam.from(theParams, specialParam);
+						double latitudeValue = location.getLatitudeValue();
+						double longitudeValue = location.getLongitudeValue();
+						final CoordsPredicateBuilder coordsPredicateBuilder = mySqlBuilder.addCoordsPredicateBuilder(resourceLinkPredicateBuilder.getColumnTargetResourceId());
+						mySqlBuilder.addSortCoordsNear(coordsPredicateBuilder, latitudeValue, longitudeValue, theAscending);
+					} else {
+						String msg = myFhirContext.getLocalizer().getMessageSanitized(QueryStack.class, "cantSortOnCoordParamWithoutValues", theParamName);
+						throw new InvalidRequestException(Msg.code(2494) + msg);
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java:395]: issue: please use error code 2497, i have reserved it for you.

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/UpliftedChainedNearSearchTest.java [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+		public String getParameterValue(FhirContext theContext) {
+			return value;
+		}
+
+		@Override
+		public String getParameterName() {
+			return name;
+		}
+	}
+}
```
Comment (tadgh) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/UpliftedChainedNearSearchTest.java:332]: Recommendation: Move these tests to `ResourceProviderR4DistanceTest`. This looks as though it is meant to test distance, just via the chain right?

Comment (tadgh) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/UpliftedChainedNearSearchTest.java:332]: Actually, looks like a similar test already exists in there: https://github.com/hapifhir/hapi-fhir/blob/5bdf4e85e20d5c113fae85e7ee0b8c1b7177c8b0/hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4DistanceTest.java#L118-L118

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/UpliftedChainedNearSearchTest.java [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+		@Override
+		public String getParameterValue(FhirContext theContext) {
+			return value;
+		}
+
+		@Override
+		public String getParameterName() {
+			return name;
+		}
+	}
```
Comment (tadgh) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/UpliftedChainedNearSearchTest.java:331]: haha, looks like we are in need of a DistanceParam

**Review by tadgh — APPROVED** [review]
Approved

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/UpliftedChainedNearSearchTest.java (hunk 3 above)
Comment (tadgh) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/UpliftedChainedNearSearchTest.java:332]: Ah I see, you are specifically checking sorting.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java (hunk 2 above)
Comment (jkiddo) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java]: Doen

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/UpliftedChainedNearSearchTest.java (hunk 3 above)
Comment (jkiddo) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/UpliftedChainedNearSearchTest.java]: Moved all tests to `ResourceProviderR4DistanceTest`

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/UpliftedChainedNearSearchTest.java (hunk 1 above)
Comment (jkiddo) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/UpliftedChainedNearSearchTest.java]: As noted in https://github.com/hapifhir/hapi-fhir/pull/5668#discussion_r1484642511 - it is chained sorting  :)

**Comment by jkiddo** [comment]
@tadgh should be all good now

**Pull request merged**